---
title: "ExprX-vignette"
author: "Ming-an Sun, Yejun Wang, Linlin Zou, Yanhua Li, Guoqiang Zhu, Todd Macfarlan"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ExprX-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## 1 Introduction
__ExprX__ is an R package to streamline interspecies differential expression
analysis. Taking TPM or FPKM/RPKM files for samples from different species as
input, it provides functions to handle all the necessary steps, including 
data loading, ortholog matching, normalization, differential analysis and 
visualization.

ExprX is freely available at: https://github.com/mingansun/ExprX.

## 2 How to install ExprX

__ExprX__ can be installed using the __install_git__ function from devtools
package. However, it depends on several other R packages, which should be 
installed first.

To install ExprX and its dependencies:
```{r, eval = FALSE}
# install several dependent packages from Bioconductor using BiocManager
install.packages("BiocManager")
BiocManager::install(c("biomaRt", "edgeR", "RankProd"))

# install ExprX from GitHub using devtools
install.packages("devtools")
devtools::install_git("https://github.com/mingansun/ExprX")
```

To load the ExprX package:
```{r setup}
library(ExprX)
```

For more details about how to install and use __ExprX__, please refer to the
website: https://github.com/mingansun/ExprX

## 3 Apply ExprX for interspecies differential expression analysis
Using RNA-Seq data for human and mouse brain, this vignette demonstrates how 
to detect differentially expressed genes between species using ExprX. All 
the involved steps are described below.

### 3.1 Generate ExprX object by integrating interspecies expression data

By parsing the meta table (as a data frame or CSV file) which contains
information about expression data files (usually contain TPM, FPKM or RPKM
values) for different species, the function __make_ExprX_dataset__ can read
these data files to create an object which contains the expression levels of
the replicates of different species. The created ExprX object can also contain
additional data such as orthologue pairs, normalized expression etc, and will
be used by most of the subsequent analysis.

To use __make_ExprX_dataset__ to read CSV file with meta data for expression
data files and compared species (ie. human and mouse) to create an ExprX object:

```{r}
# meta table file
hs2mm.meta_file <- paste0(path.package("ExprX"), "/extdata/brain_metatable.csv")

# make ExprX object from meta table
hs2mm.data  <- make_ExprX_dataset(
  hs2mm.meta_file,
  data_dir = paste0(path.package("ExprX"), "/extdata")
  )
```


### 3.2 Determine the 1-to-1 orthologs among compared species

The 1-to-1 orthologs among species are constructed based on the homolog
annotations from ENSEMBL database. Thus, only species available in
in ENSEMBL database (about 200 as checked on 2020-4-7) can be used for analysis.

To use the __list_species__ function to get the information (eg. species name, 
abbreviation) for all the species supported by ExprX:

```{r}
sp.lst <- list_species()
head(sp.lst)
```

#### Match and save 1-to-1 orthologs among species

The __ortholog_match__ function invokes __biomaRt__ package to retrieve homolog
annotation from ENSEMBL database, then matches 1-to-1 orthologs by reciprocal
comparison. This step usually takes a few minutes - depending on the network
speed). To speed up, the obtained ortholog data can be stored on hard disk with
__saveRDS__ for later use.

```{r}
# Match 1:1 orthologs between human and mouse
hs2mm.orth <- ortholog_match("human", "mouse")

# Save the ortholog results on hard disk for later use
saveRDS(hs2mm.orth, "hs2mm.orth.rds")
```

#### Load previously saved ortholog matching results from hard disk

Below shows how to used __readRDS__ to load ortholog matching result that is
previously saved on hard disk. Alternatively, the ortholog matching result can
be generated with __ortholog_match__ as demonstrated above.

The ortholog matching result includes information for each involved species,
such as GeneID, GeneName, Chrom, GeneType and so on.

```{r}
# Load ortholog result with readRDS
hs2mm.orth <- readRDS(paste0(path.package("ExprX"), "/data/hs2mm.orth.rds"))

# View the structure of hs2mm.orth
str(hs2mm.orth)
```

To summarize the ortholog results by genetype (eg. protein_coding, miRNA, lncRNA etc):

```{r}
hs2mm.orth.genetype <- summarize_ortholog_gene(hs2mm.orth, group = "genetype")
head(hs2mm.orth.genetype)
```

To summarize the ortholog matching results by chromosome:

```{r}
hs2mm.orth.chrom <- summarize_ortholog_gene(hs2mm.orth, group = "chrom")
head(hs2mm.orth.chrom)
```

#### Filter 1-to-1 orthologs to exclude specific groups of genes

In many cases, specific groups of genes (eg. pseudogenes or genes from sex chromosomes) are
undesirable for gene expression comparison. The function __ortholog_filter__ enables the
filtering of ortholog pairs based on gene type, chromosome, or provided gene list, as 
demonstrated below.

```{r}
# Filter orthologs by excluding genes from chromosomes X, Y and MT, and only keep
# protein_coding genes
hs2mm.orth.flt <- ortholog_filter(
  hs2mm.orth,
  genetype_include = "protein_coding",
  chrom_exclude = c("X", "Y", "MT")
)
```

To check the number of 1:1 orthologs before and after filtering:

```{r}
# Before filtering
dim(hs2mm.orth[[1]])

# After filtering
dim(hs2mm.orth.flt[[1]])

```

### 3.3 Integrate the 1-to-1 ortholog matching result to ExprX object

Take the original ExprX object generated with __make_ExprX_dataset__ and the
ortholog matching data generated using __ortholog_match__ as input, the function
__ortholog_expression_merge__ integrates together the expression data for all
1:1 orthologs among compared species. The integrated data will be appended to
the original ExprX object and returned as an updated object. To be noted, for
orthologs that don't have matched expression data, they will be excluded from
the returned data.

```{r}
# Merge data
hs2mm.data <- ortholog_expression_merge(
  expr_data = hs2mm.data, orth_data = hs2mm.orth.flt
)
```

### 3.4 Integrate the normalized expression data for 1-to-1 orthologs to ExprX object

Normalization of the expression data for 1:1 orthologs among species can be performed
by using the __ortholog_expression_normalize__ function. Different normalization
approaches are supported, including TMM, TMMwsp, RLE, upperquartile and quantile. The
normalized data matrix will be appended to the original ExprX object and returned as
the updated ExprX object.

To normalize the expression data of 1:1 orthologs among samples using the TMM approach:

```{r}
# Load required package
library(edgeR)

# Perform data normalization
hs2mm.data <- ortholog_expression_normalize(
  expr_data = hs2mm.data, method = "TMM"
)
```

### 3.5 Perform interspecies differential expression analysis

Differential expression analysis of 1:1 orthologs between species can be 
performed using the __ortholog_expression_compare__ function. Expression data of
1:1 orthologs after normalization are used for differential expression analysis.
Statistics such as average expression level, log2foldChange and p-values are 
calculated and returned as a dataframe. 

Below shows how to perform interspecies differential analysis for 1:1 orthologs
using RankProd approach. To be noted, the demo dataset is only for less than
1000 genes, thus only a couple of genes are called as differentlly expressed.
If use full dataset, the number of called differential genes can be as many as
several hundreds.

```{r}
# Load required package
library(RankProd)

# Perform differential analysis and then sort by p-value
hs2mm.deg <- ortholog_expression_compare(
  hs2mm.data, method = "RankProd", p_adjust = "fdr"
  )
hs2mm.deg <- hs2mm.deg[order(hs2mm.deg$P_value),]

# Check what the differential analysis result looks like
head(hs2mm.deg)
```

To determine differential genes (based on cufoff of p-value and log2foldChange)
which can be saved for downstream analysis:

```{r}
# Determine significant differential genes (human>mouse) based on p-values and log2foldChange
hs2mm.hsHigh <- subset(hs2mm.deg, subset = log2foldChange > 1  & P_value < 0.05)
head(hs2mm.hsHigh)

# Determine significant differential genes (human<mouse) based on p-values and log2foldChange
hs2mm.mmHigh <- subset(hs2mm.deg, subset = log2foldChange < -1 & P_value < 0.05)
head(hs2mm.mmHigh)
```

### 3.6 Visualize differential expression analysis result

The differential expression can be visualized as MA-plot or Volcano-plot, with
differential genes highlighted by color. Below shows how to use the
__ortholog_expression_plot__ function to generate MA-plot and Volcano-plot, 
respectively.

```{r, fig.width = 4, fig.height = 4}
# Generate MA-plot
ortholog_expression_plot(
  hs2mm.deg, "MA",
  main = "Brain", xlim = c(0,120000), ylim = c(-18,18)
  )
```

```{r, fig.width = 4, fig.height = 4}
# Generate Volcano-plot
ortholog_expression_plot(
  hs2mm.deg, "volcano",
  main = "Brain", xlim = c(-18,18), ylim = c(0, 3)
  )
```

